devlink_linux.go

```diff
@@ -0,0 +1,167 @@
+package netlink
+
+import (
+	"syscall"
+
+	"github.com/vishvananda/netlink/nl"
+	"golang.org/x/sys/unix"
+)
+
+// DevlinkDevEswitchAttr represents device's eswitch attributes
+type DevlinkDevEswitchAttr struct {
+	Mode       string
+	InlineMode string
+	EncapMode  string
+}
+
+// DevlinkDevAttrs represents device attributes
+type DevlinkDevAttrs struct {
+	Eswitch DevlinkDevEswitchAttr
+}
+
+// DevlinkDevice represents device and its attributes
+type DevlinkDevice struct {
+	BusName    string
+	DeviceName string
+	Attrs      DevlinkDevAttrs
+}
+
+func parseDevLinkDeviceList(msgs [][]byte) ([]*DevlinkDevice, error) {
+	devices := make([]*DevlinkDevice, 0, len(msgs))
+	for _, m := range msgs {
+		attrs, err := nl.ParseRouteAttr(m[nl.SizeofGenlmsg:])
+		if err != nil {
+			return nil, err
+		}
+		dev := &DevlinkDevice{}
+		if err := dev.parseAttributes(attrs); err != nil {
+			return nil, err
+		}
+		devices = append(devices, dev)
+	}
+	return devices, nil
+}
+
+func parseEswitchMode(mode uint16) string {
+	switch mode {
+	case nl.DEVLINK_ESWITCH_MODE_LEGACY:
+		return "legacy"
+	case nl.DEVLINK_ESWITCH_MODE_SWITCHDEV:
+		return "switchdev"
+	default:
+		return "unknown"
+	}
+}
+
+func parseEswitchInlineMode(inlinemode uint8) string {
+	switch inlinemode {
+	case nl.DEVLINK_ESWITCH_INLINE_MODE_NONE:
+		return "none"
+	case nl.DEVLINK_ESWITCH_INLINE_MODE_LINK:
+		return "link"
+	case nl.DEVLINK_ESWITCH_INLINE_MODE_NETWORK:
+		return "network"
+	case nl.DEVLINK_ESWITCH_INLINE_MODE_TRANSPORT:
+		return "transport"
+	default:
+		return "unknown"
+	}
+}
+
+func parseEswitchEncapMode(encapmode uint8) string {
+	switch encapmode {
+	case nl.DEVLINK_ESWITCH_ENCAP_MODE_NONE:
+		return "disable"
+	case nl.DEVLINK_ESWITCH_ENCAP_MODE_BASIC:
+		return "enable"
+	default:
+		return "unknown"
+	}
+}
+
+func (d *DevlinkDevice) parseAttributes(attrs []syscall.NetlinkRouteAttr) error {
+	for _, a := range attrs {
+		switch a.Attr.Type {
+		case nl.DEVLINK_ATTR_BUS_NAME:
+			d.BusName = string(a.Value)
+		case nl.DEVLINK_ATTR_DEV_NAME:
+			d.DeviceName = string(a.Value)
+		case nl.DEVLINK_ATTR_ESWITCH_MODE:
+			d.Attrs.Eswitch.Mode = parseEswitchMode(native.Uint16(a.Value))
+		case nl.DEVLINK_ATTR_ESWITCH_INLINE_MODE:
+			d.Attrs.Eswitch.InlineMode = parseEswitchInlineMode(uint8(a.Value[0]))
+		case nl.DEVLINK_ATTR_ESWITCH_ENCAP_MODE:
+			d.Attrs.Eswitch.EncapMode = parseEswitchEncapMode(uint8(a.Value[0]))
+		}
+	}
+	return nil
+}
+
+func parseEswitchAttrs(msgs [][]byte, dev *DevlinkDevice) {
```

Comment:
May be make that func a method of *DevlinkDevice like parseAttributes?
And may be return an error? Or it should be silent?

Reply:
@callmehollow now made it a method of *DevlinkDevice.